token-metadata/program/src/state.rs

```diff
@@ -0,0 +1,38 @@
+use {
+    borsh::{BorshDeserialize, BorshSerialize},
+    solana_program::pubkey::Pubkey,
+};
+/// prefix used for PDAs to avoid certain collision attacks
+pub const PREFIX: &str = "metadata";
+
+pub const NAME_LENGTH: usize = 32;
+
+pub const SYMBOL_LENGTH: usize = 10;
+
+pub const URI_LENGTH: usize = 200;
+
+pub const METADATA_LEN: usize = 32 + NAME_LENGTH + SYMBOL_LENGTH + URI_LENGTH + 200;
+
+pub const OWNER_LEN: usize = 32 + 32 + 200;
+
+#[repr(C)]
+#[derive(Clone, Default, BorshSerialize, BorshDeserialize)]
+pub struct Metadata {
+    /// Mint of the token asset
+    pub mint: Pubkey,
+    /// The name of the asset
+    pub name: String,
+    /// The symbol for the asset, ie, AAPL or SHOES
+    pub symbol: String,
+    /// URI pointing to JSON representing the asset
+    pub uri: String,
+}
+
+#[repr(C)]
+#[derive(Clone, Debug, Default, PartialEq, BorshSerialize, BorshDeserialize)]
+pub struct Owner {
```

Comment:
```suggestion
pub struct UpdateAuthority {
```

Reply:
Actually why not just put the owner address in the Metadata account and avoid the Owner account entirely

Reply:
Because I want to ensure that any name/symbol combo can only be taken once within the scope of a given program, so by creating a separate account indexed by [program_id, name, symbol] where the owner key lies, we ensure that only one key gets to own that name/symbol combo. One can look up metadata on a mint, find it's name and symbol if there is one, and if it exists and they want to know who owns that mint, they can find it out by doing one further lookup on [program_id, name, symbol.] This maintains compatibility ERC-721 - that's why owner is not included with object, but in a separate account where it belongs. Ultimately, owner is not metadata. But yeah, we can change it from owner to "authority" as far as naming goes.

Reply:
Why not include `mint address` in  the seeds in addition to `[program_id, name, symbol]` to avoid this problem?

Reply:
Because then you could still have multiple accounts made for the same name/symbol combo, because one person could make a NIKE/SWOOSH combo with mint A and another could make Mint B and make NIKE/SWOOSH as well and get a separate account. In order for the NIKE brand to get away with getting their unique owner account, we need to not put Mint ID in the seed, because it ruins the uniqueness.

Reply:
If the mint address is not in the seed, it seems like I could front-run.   Like the first thing I'd do is park on common symbol names and prevent the real owners from using them.   That is the program_id is not enough to guarantee uniqueness, the mint address is also needed.  Actually the mint address is probably sufficient, since that also identifies the program id in use: `[program_id, name, symbol]` -> `[mint_address, name, symbol]` probably would work great

Reply:
You could front run, yes, but I think the greater danger is in people proliferating false clones and creating mirages of true NFTs to draw away capital. You really have a trade off situation here, and I think the better option is to allow the front running which would create an interesting market for people selling NAME/SYMBOL accounts (Which themselves become NFTs) and keep false clones from occurring, rather than the the alternative where users have no idea who the real McCoy is.

Reply:
> create an interesting market for people selling NAME/SYMBOL accounts (Which themselves become NFTs) 

Building systems that allow people to take advantage of other people doesn't feel right.  

> and keep false clones from occurring, rather than the the alternative where users have no idea who the real McCoy is.

Well no, the mint address already uniquely identifies the real McCoy.   If I'm USDC or some other 🐳 token and somebody has already camped on my desired symbol, I'm not going to negotiate with the terrorist.  I'll probably just declare the standard broken and walk away.

But there's another angle to consider here: in the future there will likely be multiple SPL Token program ids.

Reply:
Ultimately, we're trying to protect the users. Yes, the mint is unique. At the same time, all the applications the users are using depend on Name/Symbol. Just like while IP is unique, all the browsers that people use depend on the URL. This is an analogous situation.

In terms of other token programs existing - once that exists, we could add support reusing the name and symbol across multiple mints, so the person can control the name and symbol across many mints. 

At the end, there will be potentially a market for these names and symbols where people can trade them on an order book, themselves as NFTs, potentially using the metadata program itself to wrap them. That's not taking advantage of people, I don't think, these names will be trading at fair market value.

Reply:
Wait what are we talking about again?  I think there's two things here?

1.  The owner account doesn't need to exist.  Instead the "update authority" address can just be in the Metadata account.  This doesn't seem to be contested.

2. If I create token X and can't name my token X because somebody else also has a token X, I'm just going to walk away.  I'm not going to find some crazy defi derivative market for token names in an attempt to purchase the right to name my token X.   Perhaps somebody will want to build this one day, but for now I see that as completely out of scope.   I just want to assign the exact name/symbol/metadata to my token that I need, so that it shows up in the explorer and elsewhere exactly how I want it to.   To this end, using the `mint_address` alone as the program-derived address for the metadata is even better,  no need to put the symbol/name in there at all.   I have a mint address and I need to lookup the name/symbol/metadata